### ISSUE TITLE
s3_bucket crashes RStudio in 15.0.1

### ISSUE THREAD

**Issue by joranE** [apache/arrow#40678]
### Describe the bug, including details regarding any error messages, version, and platform.

Any attempt at instantiating an S3Filesystem object in 15.0.1 crashes R in RStudio. You get the dialog box that the R session has crashed and your only option is to restart the session. There are no error messages in the console.

Rolling back to 14.0.0.2 resolves the issue.

I am experiencing this on an M1 Mac running Sonoma 14.2.1.

### Component(s)

R

**Comment by blongworth** [comment]
Have you checked that arrow was compiled with S3 support with `arrow_info()`? The most recent version on CRAN has issues with the arrow binary, discussing in #40667.

**Comment by joranE** [comment]
Yes, sorry, that was the first thing I checked. They report the following:

```
> arrow_info()
Arrow package version: 15.0.1

Capabilities:
               
acero      TRUE
dataset    TRUE
substrait FALSE
parquet    TRUE
json       TRUE
s3         TRUE
gcs       FALSE
utf8proc   TRUE
re2        TRUE
snappy     TRUE
gzip       TRUE
brotli     TRUE
zstd       TRUE
lz4        TRUE
lz4_frame  TRUE
lzo       FALSE
bz2        TRUE
jemalloc   TRUE
mimalloc  FALSE

Memory:
                  
Allocator jemalloc
Current    0 bytes
Max        0 bytes

Runtime:
                        
SIMD Level          none
Detected SIMD Level none

Build:
                                    
C++ Library Version           15.0.0
C++ Compiler              AppleClang
C++ Compiler Version 15.0.0.15000100

> arrow_with_s3()
[1] TRUE
```

I was a little surprised that the binary wasn't built with gcs support, but it's reporting s3 support. When I run `s3_bucket` it appears to run through the auth process (for out setup, I'm redirected to a web page, verify an alphanumeric code, click authorize) and then the R session immediately crashes.

**Comment by amoeba** [comment]
Thanks for the report @joranE. How did you install arrow here? I'd like to try to reproduce the error so we can start debugging your issue. If you just ran `install.packages`, I'm guessing something about your setup triggered a source build. If you're not sure, running the following and pasting the output here would be helpful:

```r
Sys.setenv("ARROW_R_DEV"=TRUE)
install.packages("arrow")
```

**Comment by joranE** [comment]
You're right, sorry again, it was the end of a long workday I'm usually much better about bug reports.

My initial install was in the context of an renv project and you're right it was pulling from the Posit Package Manager (`https://packagemanager.posit.co/cran/latest`) which doesn't have macOS binaries so it built 15.0.1 from source on my machine. That's the version from which I was originally reporting, with the apparent presence of S3 support but R session  crashes upon use of `s3_bucket`. Reading & writing local parquet files worked fine with that version built from source.

Subsequently, when I've tried installing the binary from CRAN (`https://cran.rstudio.com`) I see what folks have reported from the other issue, i.e. `arrow_info()` reports essentially zero capabilities as below:

```
> arrow_info()
Arrow package version: 15.0.1

Capabilities:
               
acero      TRUE
dataset   FALSE
substrait FALSE
parquet   FALSE
json      FALSE
s3        FALSE
gcs       FALSE
utf8proc  FALSE
re2       FALSE
snappy    FALSE
gzip      FALSE
brotli    FALSE
zstd      FALSE
lz4       FALSE
lz4_frame FALSE
lzo       FALSE
bz2       FALSE
jemalloc  FALSE
mimalloc  FALSE

Memory:
                 
Allocator  system
Current   0 bytes
Max       0 bytes

Runtime:
                        
SIMD Level          none
Detected SIMD Level none

Build:
                                    
C++ Library Version           15.0.1
C++ Compiler              AppleClang
C++ Compiler Version 14.0.0.14000029
```
